### PR TITLE
Tigers - Renee Su, Yael Echols

### PR DESCRIPTION
**Pull request by reneesu99**

### CODE REVIEW

File: adagrams/game.py [hunk 1]
```diff
@@ -1,11 +1,64 @@
...
+    }
+
+    score = [val for letter in word.upper() for key, val in scores.items() if letter in key]
+
+    if len(score) >= 7:
+        score.append(8)
+
+    return sum(score)
 
 def get_highest_word_score(word_list):
```
Comment (yaelso) [inline adagrams/game.py:44]: Just wanted to add some commentary to this get_highest_word_score() func! I initially considered utilizing a word:score dictionary here, but settled on the tuples because of ease and clarity in accessing/sorting each word. I realize my solution might not be as efficient as a dict that only goes through one score sort and one tie breaker sort, however, and I've played around with a few varying implementations since submission to try and maximize efficiency thereby!

File: README.md [hunk 2]
```diff
@@ -1,5 +1,22 @@
...
+
+Come together to go over waves 1-3 and pair program wave 4
+Wednesday
+
+
+Communication:
+Communicate via Slack for waves 1-3 unless video support needed
+
+
+ -->
```
Comment (chimerror) [inline README.md:19]: Ooh, I like this planning!

File: README.md [hunk 3]
```diff
@@ -290,4 +309,8 @@ Implement a function called `get_highest_word_score` in `game.py`. This method s
 
 
 
-
+Returns
+Wave 1 draw_letters: array of 10 strings ["d","e","f" ... ]
+Wave 2 use_available_letters: expecting boolean based on validity of input (word and letter bank)
+Wave 3 score_word: returns integer value corresponding to points a word gets
+Wave 4 get_highest_word_score: returns a tuple that represents winning word and score
```
Comment (chimerror) [inline README.md:316]: Ah, definitely a good habit to write down clarifications of requirements after you get them. Memory is very fickle...

File: adagrams/game.py [hunk 4]
```diff
@@ -1,11 +1,64 @@
+from random import sample
```
Comment (chimerror) [inline adagrams/game.py:1]: Very minor style nitpick: don't be afraid to add blank lines to provide some separation between sections of your code! I think one here after this `import` and before `draw_letters` would be nice.

File: adagrams/game.py [hunk 5]
```diff
@@ -1,11 +1,64 @@
...
 def draw_letters():
```
Comment (chimerror) [inline adagrams/game.py:2]: This is a very succinct way to write this function and very good use library functions, which is great! That being said, I do have some comments, more advisory than anything...

File: adagrams/game.py [hunk 6]
```diff
@@ -1,11 +1,64 @@
...
-    pass
+    hand = sample(['A','B','C','D','E','F','G','H','I','J','K','L',\
```
Comment (chimerror) [inline adagrams/game.py:3]: While making sure to mark line continuations using `\` is exactly right with in Python, line breaks _are_ acceptable within braces, so y'all don't need them here, since y'all are in the list definition braces (`[]`).

File: adagrams/game.py [hunk 7]
```diff
@@ -1,11 +1,64 @@
...
+                   'M','N','O','P','Q','R','S','T','U','V','W','X',\
+                   'Y','Z'], counts = [9,2,2,4,12,2,3,2,9,1,1,4,2,6,\
+                                       8,2,1,6,4,6,4,2,2,1,2,1], k = 10)
```
Comment (chimerror) [inline adagrams/game.py:6]: Being succinct is very good, and y'all do a great job of it here by using `sample`. However, there are readability and maintainability downsides to being so succinct like this.

Mainly, consider how hard it would be for a reader to tell which tile count corresponds to each letter from this. This also would affect the ease at which a future editor would select the right count to update if they were trying to implement a feature change to a different tile distribution. For example, let's say I (poor sap who inherited y'all's code) wanted to increase the number of "S" tiles from 4 to 5. I basically would need to count through the counts to find the right `4` entry to update, which can be error-prone. (I did find it but I had to say my ABCs to myself, lol.) 

There are a few ways we could rearrange the code to counter this. One way might be to put the lists into variables, keeping them parallel:

```python
LETTERS = [
    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
    'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'
]
LETTER_COUNTS = [
    9, 2, 2, 4, 12, 2, 3, 2, 9, 1, 1, 4, 2,
    6, 8, 2, 1, 6, 4, 6, 4, 2, 2, 1, 2, 1
]
hand = sample(LETTERS, counts=LETTER_COUNTS, k=10)
return hand
```

(Note that I also capitalized the variable names as that is the Python convention for marking constants like these, and changed the spacing a bit.)

But even that feels tricky to me, so my honest preference would be to use a dictionary:

```python
TILE_COUNTS = {
    'A': 9,
    'B': 2,
    'C': 2,
    'D': 4,
    # and so on...
}
hand = sample(TILE_COUNTS.keys(), counts=TILE_COUNTS.values(), k=10)
return hand
```

Admittedly that makes for a very long dictionary (which is why I didn't include the whole one in my snippet above), but I think that's worth having a clear tie between the letters and their tile counts. Updating the number of "C" tiles from 2 to 3 for example would be much easier this way.

File: adagrams/game.py [hunk 8]
```diff
@@ -1,11 +1,64 @@
...
+    return hand
 
 def uses_available_letters(word, letter_bank):
```
Comment (chimerror) [inline adagrams/game.py:9]: This looks good! No comments!

File: adagrams/game.py [hunk 9]
```diff
@@ -1,11 +1,64 @@
...
+        if letter not in letter_bank_dict:
+            return False
+        elif letter_bank_dict[letter] == 0:
+            return False
+        else:
+            letter_bank_dict[letter] -= 1
+    return True
+
 
 def score_word(word):
```
Comment (chimerror) [inline adagrams/game.py:26]: This is also very succinct and a very good use of list comprehensions, but here the desire to be succinct actually causes a performance hit both in time and space! (I also have a naming comment...)

File: adagrams/game.py [hunk 10]
```diff
@@ -1,11 +1,64 @@
...
+        "AEIOULNRST" : 1,
+        "DG" : 2,
+        "BCMP" : 3,
+        "FHVWY" : 4,
+        "K" : 5,
+        "JX" : 8,
+        "QZ" : 10
+    }
+
+    score = [val for letter in word.upper() for key, val in scores.items() if letter in key]
```
Comment (chimerror) [inline adagrams/game.py:37]: Very minor naming issue: Since this generates a list, a singular name like `score` is very misleading over a plural name like `scores`. However, since y'all already used `scores` above, I get the conflict. Even then a name like `letter_scores` or something like that could work.

However, `scores` itself is a constant and probably should be written `SCORES`, though I would probably choose a name like `TILE_POINTS`, which then frees us up to use `scores` for this list...

Comment (chimerror) [inline adagrams/game.py:37]: But perhaps more major, the use of a list comprehension as well with the design of the `scores` dictionary causes a time performance issue that might be clearer to see if I expand the comprehension out into explicit loops:

```python
score = []
for letter in word.upper():
    for key, val in scores.items():
        if letter in key:
            score.append(val)
```

(I maintained y'all's names just so the tie to y'all's code is clearer.)

To demonstrate the issue, I will have to be a bit more exact than jumping to Big-O, so let's let n be the number of letters in `word`. For each letter in `word`, y'all's implementation will possibly go through every key-value pair in `scores`. But in addition to that, keep in mind that `letter in key` will have to possibly go through every letter in `key` because that is how `in` works with a string or list. That's a lot of extra work!

If we think of the worst case, the letter "Z", y'all's implementation will have to go through every other letter in all the keys before it gets to "Z". So let's just use 26. So doing 26 operations n times puts my calculation here at 26n operations. Which is O(n), and not that bad, but I want y'all to focus on 26n here.

If instead we used a "flat" dictionary for `scores` (which I'm going to keep calling `scores` just so the tie to y'all's code is clear, as well as use a comprehension):

```python
scores = {
    'A': 1,
    'B': 3,
    'C': 3,
    # and so on...
}
score = [scores[letter] for letter in word.upper()]
```

Now we don't need to go through all the keys in `scores`! We can access the values directly, and dictionary access is O(1), so I'm going to only treat that as 1 so this is just n operations.

Of course n and 26n are both O(n), but doing 26 times more work is still rough, so it's worth avoiding! (Consider this an example of how there's more to performance than just Big-O...)

Comment (chimerror) [inline adagrams/game.py:37]: In addition to the time complexity issue, there is also a space complexity issue by using a comprehension, which exists regardless of the data structure of `scores`.

Namely, `score` will have up to n + 1 elements, n for each letter in `word`, and possibly one more for the long word bonus. This makes space complexity O(n). But we could just use a variable to keep the score as well (which would make much more sense to be named `score`):

```python
score = 0
for letter in word.upper():
    score += scores[letter]

if len(word) >= 7:
    score += 8

return score
```

(I also went ahead and assumed that `scores` was my flat dictionary from the last comment because it's less to write...)

Sadly, we do lose the cool comprehension, but in this case no extra space is allocated at all! We have `scores` and `score`, but both of those are constant in space, and do not change as the number of letters in `word` changes. Making this implementation O(1) in space complexity, which _is_ a big enough difference to change Big-O!

File: adagrams/game.py (hunk 1 above)
Comment (chimerror) [reply adagrams/game.py:44]: Actually in this case, I think the list of tuples are fine! No matter what, you'll have to go through either all the tuples or all the key-value pairs of a dictionary to determine the highest score and perform tiebreakers. Additionally, given the tiebreaker that prefers the word that occurs first in `word_list`, maintaining that order makes implementing that tiebreaker easier!

A dictionary would help if we could somehow create it without going through all the words in `word_list`. For example that's why it would help in `score_word` above, since that is a constant dictionary. As I said in my comments there, it allows us to access the values directly rather than iterating through them, which is the big benefit of dictionaries over lists. But by the time we would create such a dictionary for `get_highest_word_score`, it's too late...

Comment (chimerror) [inline adagrams/game.py:44]: This is a good working, readable implementation, but I do see a performance issue due to using `sort`, as well as a logic error in the tiebreaking rules...

File: adagrams/game.py [hunk 11]
```diff
@@ -1,11 +1,64 @@
...
+        if word[1] == highest_scoring[0][1]:
+            highest_scoring.append(word)
+        elif word[1] > highest_scoring[0][1]:
+            highest_scoring.clear()
+            highest_scoring.append(word)
+
+    if len(highest_scoring) == 1:
+        return highest_scoring[0]
+
+    highest_scoring.sort(key=lambda word: len(word[0]))
```
Comment (chimerror) [inline adagrams/game.py:59]: So while this is a great and correct use of `sort`, note that `sort` doesn't come for free! Python's implementation of `sort` takes O(n * log(n)) time. While most likely there'll be a lot less words in `highest_scoring` than `word_list`, if every word in `word_list` has the same score, then they will all be in `highest_score`, bringing the total time complexity here to O(n * log(n)), where n is the number of words in `word_list`. However it can be done in O(n) by not bothering to keep track of previous highest scoring words and performing the tie-breaking within the for above loop (I'm also going to make a few other nice code changes that I'll point out in comments in the snippet):

```python
for word_score_tuple in score_list:
    # using tuple destructuring to give the tuple elements some nicer names
    word, score = word_score_tuple
    highest_word, highest_score = highest_scoring
    if score == highest_score:
        highest_len = len(highest_word) # only using a variable here and with word_len for a nicer name
        if highest_len != 10:
            word_len = len(word)
            if word_len == 10 or word_len < highest_len:
                 highest_scoring = word_score_tuple
    elif score > highest_score:
         highest_scoring = word_score_tuple

return highest_scoring
```

Yikes! That is a lot of nesting, which hurts readability, but with the benefit that we only have to go through `score_list` once! That's O(n), which is a big improvement over O(n * log(n))!

File: adagrams/game.py [hunk 12]
```diff
@@ -1,11 +1,64 @@
...
+
+    if len(highest_scoring[-1][0]) == 10:
+        highest_scoring.sort(key=lambda word: len(word[0]), reverse=True)
```
Comment (chimerror) [inline adagrams/game.py:62]: While this `if` block will usually select a 10-letter word as a tiebreaker, and we can assume no word uses more than 10 letters, and thus the 10 letter words are at the end of `highest_scoring`, there is a logic error if there are _multiple_ 10-letter words that causes this implementation to not match the requirements.

From the project requirements: "If the there are multiple words that are the same score and the same length, pick the first one in the supplied list". Given that, with two 10-letter highest scoring words, we would want to pick whichever word was given first in `word_list`. However, y'all's code is picking the _last_ word in `highest_scoring` (if it has 10 letters), which will _not_ be the 10-letter word that came first in `word_list` but the 10-letter word that came _last_.

My suggested `for` loop in above comment avoids this error, obviously...

Comment (chimerror) [reply adagrams/game.py:62]: You may be wondering how I knew that the last word in `highest_scoring` was also the last relevant word in `word_list` despite the call to `highest_scoring.sort` on line 59. Basically the `sort` function in Python is guaranteed to be a stable sort. That means when there are ties, the order from the original list is maintained.

So let's go through the full genealogy of `highest_scoring`. The list comprehension that sets `score_list` will maintain the order of `word_list`. We then go through `score_list` to append to `highest_scoring`, so that order will be maintained. And then since `sort` is a stable sort, even calling that will maintain the order among ties.

**Review by chimerror — COMMENTED** [review]
Good work!

This implementation is correct and very succinctly written while generally being readable. I did make some comments about some of y'all's data structure choices, which impacts readability, maintainability, and even causes small performance issues. I also made comments about naming, as well as a performance issue and logic error in `get_highest_word_score`. Despite that, I think this is good enough to be marked Green!

File: adagrams/game.py (hunk 10 above)
Comment (chimerror) [reply adagrams/game.py:37]: I'll point out that this comment about space complexity does _not_ apply if rather than a comprehension, you use a generator (working from my improved "flat" dictionary implementation):

```python
score = sum(scores[letter] for letter in word.upper())
```

Note that rather than having square brackets, I simply put the same expression into a call to sum. When it is like this, rather than a list comprehension it becomes a "generator". Which is essentially equivalent to this function:

```python
def score_letters(scores, word):
    for letter in word.upper():
        yield scores[letter]
```

And then you can imagine it as if `sum` was written like this:

```python
def sum(generator):
    total_sum = 0
    for value in generator:
        total_sum += value
    return sum
```

So basically my original use of the generator would become `score = sum(score_letters(scores, word))`, and this ends up doing the exact same thing as in my suggestion, just divided across two functions.

Here's a nice article on generators, because they are a very useful part of Python (and other languages have similar features): https://realpython.com/introduction-to-python-generators/

Comment (chimerror) [reply adagrams/game.py:37]: In fact, you can think of a comprehension as just a shorthand way of actually allocating the results of a generator expression into a data structure!

So `[scores[letter] for letter in word.upper()]` is like:

```python
def score_letters_comprehension(scores, word):
    generator = scores[letter] for letter in word.upper()
    score = []
    for val in generator:
        score.append(val)
    return score
```
